### PR TITLE
Protect Routes

### PR DESCRIPTION
**Pull request by MarioRodrigues10**

### CODE REVIEW

**Review by ruioliveira02 — CHANGES_REQUESTED** [review]
Whenever I try to create a:
- Company
- Job
- Recruiter (though for a different reason: name used to be company and hasn't been properly updated)
- FAQS

the page is crashing, as it is being directed to a `show`, despite having no ID. Print:
image

**Review by ruioliveira02 — COMMENTED** [review]
image
Fix this error pls :pleading_face: 

image
 Also this

**Review by ruioliveira02 — CHANGES_REQUESTED** [review]
I am worried about two things:

1. I am not seeing any distinctions between the non admin roles (staff, attendee and recruiter). 
2. Staff has way too much access. The only creative action (meaning performing an action that changes the state of the database) is redeeming tasks and products.

File: lib/parzival_web/live/app/announcement_live/index.html.heex [hunk 1]
```diff
@@ -2,7 +2,7 @@
   <div class="pt-5 pr-6 pb-5 pl-4 border-gray-200 sm:pl-6 lg:pl-8 xl:pt-6 xl:pl-6 border-y">
     <div class="flex justify-between items-center h-10">
       <h1 class="flex-1 text-lg font-medium">Announcements</h1>
-      <%= if @current_user.role in [:admin] do %>
+      <%= if @current_user.role in [:admin, :staff] do %>
```
Comment (RuiL1904) [inline lib/parzival_web/live/app/announcement_live/index.html.heex:5]: Only an admin should make an annoucement

File: lib/parzival_web/live/app/announcement_live/show.html.heex [hunk 2]
```diff
@@ -2,7 +2,7 @@
   <div class="pt-5 pr-6 pb-5 pl-4 border-gray-200 sm:pl-6 lg:pl-8 xl:pt-6 xl:pl-6 border-y">
     <div class="flex justify-between items-center h-10">
       <h1 class="flex-1 text-lg font-medium">Announcement</h1>
-      <%= if @current_user.role in [:admin] do %>
+      <%= if @current_user.role in [:admin, :staff] do %>
```
Comment (RuiL1904) [inline lib/parzival_web/live/app/announcement_live/show.html.heex:5]: Only an admin should edit an annoucement

File: lib/parzival_web/live/app/gamification/mission_live/index.html.heex [hunk 3]
```diff
@@ -3,7 +3,7 @@
     <div class="flex justify-between items-center h-10">
       <h1 class="flex-1 text-lg font-medium">Missions</h1>
       <!-- Action buttons -->
-      <%= if @current_user.role in [:admin] do %>
+      <%= if @current_user.role in [:admin, :staff] do %>
```
Comment (RuiL1904) [inline lib/parzival_web/live/app/gamification/mission_live/index.html.heex:6]: Only an admin should create a mission

File: lib/parzival_web/live/app/gamification/mission_live/show.html.heex [hunk 4]
```diff
@@ -3,7 +3,7 @@
     <div class="flex justify-between items-center h-10">
       <h1 class="flex-1 text-lg font-medium">Mission</h1>
       <!-- Action buttons -->
-      <%= if @current_user.role in [:admin] do %>
+      <%= if @current_user.role in [:admin, :staff] do %>
```
Comment (RuiL1904) [inline lib/parzival_web/live/app/gamification/mission_live/show.html.heex:6]: Only an admin should edit a mission

**Review by ruioliveira02 — CHANGES_REQUESTED** [review]
## Bugs

image
Route: `http://localhost:4000/app/offers/<uuid>/edit`

image
When accessing this route as an attendee (needless to say an attendee shouldn't even have access to this). Same for `companies_new/new`

image
http://localhost:4000/app/companies/15748dc5-9a74-4e3f-bdc9-50640512b669/edit as attendee. Same for profile edit

## Authorization problems

image
Staff cannot edit difficulties

image
This one is debatable, but I think staff should not have access to this

Why isn't there any link for the scanner route for staff? That is the single most important page for them

Attendees are not able click tasks as to get the QR code to redeem them.

Recruiters cannot edit offers

Staff and recruiters are being redirected to `/app` if they enter the url `/admin/scanner` manually

**Comment by MarioRodrigues10** [comment]
> Attendees are not able click tasks as to get the QR code to redeem them.

image

**Comment by MarioRodrigues10** [comment]
> When accessing this route as an attendee (needless to say an attendee shouldn't even have access to this). Same for companies_new/new
> http://localhost:4000/app/companies/15748dc5-9a74-4e3f-bdc9-50640512b669/edit as attendee. Same for profile edit


In production they will receive a 500 page

**Comment by ruioliveira02** [comment]
> > When accessing this route as an attendee (needless to say an attendee shouldn't even have access to this). Same for companies_new/new
> 
> > http://localhost:4000/app/companies/15748dc5-9a74-4e3f-bdc9-50640512b669/edit as attendee. Same for profile edit
> 
> In production they will receive a 500 page

Doesn't matter. You should never return a 500. 500 means something you were not expecting happen, and is an error situation. This is not an error situation, this is a lack of authorization problem. They should not even have access to that route

**Review by ruioliveira02 — APPROVED** [review]

**Pull request merged**